### PR TITLE
SALTO 4242 (Okta): support additions of app user schema

### PR DESCRIPTION
**Pull request by SeggevHaimovich**
_This PR will make Salto support additions of appUserSchema._
When adding Application, the client creates automatically the related appUserSchema.
We behave as if the user added the appUserSchema but behind the scenes we fetch the auto-created appUserSchema and modifying it.

In addition, changes in the base field of appUserSchema are unsupported with Okta API, therefore we warn about these changes and ignore them.

Notes:
* **We are currently ignore the option to change the `required` field within base, we may change it in the future**
* When modifying the base field, the user will get a warning that he cannot do it and that other changes will be deployed. In his env after the deployment he will see the element as if the base have changed. After fetch, he will see the real base field.

---

_Additional context for reviewer_
Jira ticket: https://salto-io.atlassian.net/browse/SALTO-4242

---
_Release Notes_:
Okta Adapter:
* support additions of app user schema

---
_User Notifications_: 
_None_

### CODE REVIEW

**Review by idoamit8 — COMMENTED** [review]
I haven't looked at the code yet, but I have two quick high-level questions:

Why are there separate filters for removal and for modifying/adding? Having two different filters makes maintenance harder.

Regarding this point:

> When modifying the base field, the user will get a warning that they cannot do it and that other changes will be deployed. In their environment after the deployment, they will see the element as if the base has changed. After fetching, they will see the real base field.


I don't understand the reason for this inconsistency between deployment and fetching. Why does this happen? We can modify the instance in onDeploy if necessary.

File: packages/okta-adapter/src/change_validators/app_user_schema_base_changes.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  ChangeError,
+} from '@salto-io/adapter-api'
+import { values } from '@salto-io/lowerdash'
+import _ from 'lodash'
+import { APP_USER_SCHEMA_TYPE_NAME } from '../constants'
+
+const getErrorFromChange = (
+  change: ModificationChange<InstanceElement> | AdditionChange<InstanceElement>,
+): ChangeError | undefined => {
+  if (isAdditionChange(change)) {
```
Comment (idoamit8) [inline packages/okta-adapter/src/change_validators/app_user_schema_base_changes.ts:34]: Since this message will be visible to users with every addition, we may want to reduce its severity to an informational level. Additionally, it would be beneficial to have a Product Manager review the message for clarity and appropriateness.

File: packages/okta-adapter/src/change_validators/app_user_schema_base_changes.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  const afterWithSameBase = after.clone()
+
+  if (_.isUndefined(beforeBase)) {
+    _.unset(afterWithSameBase.value, ['definitions', 'base'])
+  } else {
+    _.set(afterWithSameBase.value, ['definitions', 'base'], beforeBase)
+  }
+
+  if (!_.isEqual(beforeBase, afterBase)) {
+    if (_.isEqual(afterWithSameBase.value, change.data.before.value)) {
```
Comment (idoamit8) [inline packages/okta-adapter/src/change_validators/app_user_schema_base_changes.ts:56]:
```suggestion
if (_.isEqual(afterWithSameBase.value, before.value)) {
```

File: packages/okta-adapter/src/change_validators/app_user_schema_base_changes.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+  const afterWithSameBase = after.clone()
+
+  if (_.isUndefined(beforeBase)) {
+    _.unset(afterWithSameBase.value, ['definitions', 'base'])
+  } else {
+    _.set(afterWithSameBase.value, ['definitions', 'base'], beforeBase)
+  }
+
+  if (!_.isEqual(beforeBase, afterBase)) {
```
Comment (idoamit8) [inline packages/okta-adapter/src/change_validators/app_user_schema_base_changes.ts:55]: nit: Could you check this condition first?

```
if (_.isEqual(beforeBase, afterBase)) {
  return undefined;
}
```
Then proceed with the logic if they are not equal. This will make the code cleaner and reduce indentation.

File: packages/okta-adapter/src/config.ts [hunk 4]
```diff
@@ -505,11 +505,15 @@ const DEFAULT_TYPE_CUSTOMIZATIONS: OktaSwaggerApiConfig['types'] = {
       serviceUrl: '/admin/universaldirectory#app/{_parent.0.id}',
     },
     deployRequests: {
-      // Hack to pass through createCheckDeploymentBasedOnConfigValidator validator for removals
+      // Hack to pass through createCheckDeploymentBasedOnConfigValidator validator for removals and additions
```
Comment (idoamit8) [inline packages/okta-adapter/src/config.ts:508]: Lets keep the order of add, modify, remove

**Review by idoamit8 — COMMENTED** [review]
Great work! 💯 I have added a few comments.

File: packages/okta-adapter/src/change_validators/app_user_schema_base_changes.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+): ChangeError | undefined => {
+  if (isAdditionChange(change)) {
+    return {
+      elemID: change.data.after.elemID,
+      severity: 'Warning',
+      message: 'Cannot deploy the base field in App User Schema',
+      detailedMessage:
+        'Okta API does not support deploying the base field in App User Schema. Salto will deploy the element without the base field. After fetch the field will be updated as the client',
+    }
+  }
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/change_validators/app_user_schema_base_changes.ts:42]: I'm not sure if we should have 3 different messages here.. cause the bottom line is that we don't support changes to the base attributes (also change terminology to "base attributes" and not "base field" and this is the terms in the UI). 

therefore, I would phrase something in the spirit of  -
message: `Base attributes will not be deployed`
detailedMessage: `Salto cannot make changes to base attributes, as they are determined by the associated application.`

File: packages/okta-adapter/src/change_validators/app_user_schema_base_changes.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  const beforeBase = before.value.definitions?.base
+  const afterBase = after.value.definitions?.base
+
+  const afterWithSameBase = after.clone()
+
+  if (_.isUndefined(beforeBase)) {
+    _.unset(afterWithSameBase.value, ['definitions', 'base'])
+  } else {
+    _.set(afterWithSameBase.value, ['definitions', 'base'], beforeBase)
+  }
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/change_validators/app_user_schema_base_changes.ts:53]: why is this needed?

**Review by shir-reifenberg — COMMENTED** [review]
Looks good (only did partial review, will leave @idoamit8 to continue :) )
I think we should changes messages a bit, we can also discuss IRL

File: packages/okta-adapter/src/filters/app_user_schema_deployment.ts [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+ * Deploy changes of appUserSchema.
+ * additions - changing them to modification changes,
+ * because appUserSchema automatically created by the service when deploying new Application
+ * modifications - changing the base field to the original value because Okta's API doesn't support changing it
+ * removals - verifying the parent application is deleted. appUserSchema is deleted if and only if the parent application is deleted
+ */
+const filterCreator: FilterCreator = ({ definitions, oldApiDefinitions }) => ({
+  name: 'appUserSchemaDeployment',
+  deploy: async changes => {
+    const client = definitions.clients.options.main.httpClient as OktaClient
```
Comment (idoamit8) [inline packages/okta-adapter/src/filters/app_user_schema_deployment.ts:187]:
```suggestion
const client = definitions.clients.options.main.httpClient
```

File: packages/okta-adapter/src/filters/app_user_schema_deployment.ts [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+  const appUserSchemaInstance = getChangeData(change)
+  const parentApplicationId = getParents(appUserSchemaInstance)[0]?.id
+  if (!_.isString(parentApplicationId) || !(await verifyApplicationIsDeleted(parentApplicationId, client))) {
+    throw new Error('Expected the parent Application to be deleted')
+  }
+}
+
+const deployChange = async (
+  change: Change<InstanceElement>,
+  client: OktaClient,
```
Comment (idoamit8) [inline packages/okta-adapter/src/filters/app_user_schema_deployment.ts:165]:
```suggestion
client: clientUtils.HTTPWriteClientInterface & clientUtils.HTTPReadClientInterface,
```

Also in all other functions in this file.

File: packages/okta-adapter/test/change_validators/app_user_schema_base_changes.test.ts [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+  beforeEach(() => {
+    jest.clearAllMocks()
+    appUserSchemaInstance = appUserSchema.clone()
+  })
+  describe('addition changes', () => {
+    it('should return info when adding an appUserSchema', async () => {
+      const changes = [toChange({ after: appUserSchemaInstance })]
+      const changeErrors = await appUserSchemaBaseChangesValidator(changes)
+      expect(changeErrors).toHaveLength(1)
+      expect(changeErrors[0]).toEqual(getErrorWithSeverity(appUserSchemaInstance.elemID, 'Info'))
```
Comment (idoamit8) [inline packages/okta-adapter/test/change_validators/app_user_schema_base_changes.test.ts:74]: When we write tests, we usually write the text as-is, without calling functions or using ${}. This acts as a safeguard to verify there are no spelling mistakes, and that the text appears as expected.

**Review by idoamit8 — COMMENTED** [review]
Great work! 💯
I've added some small comments.

@shir-reifenberg  I see it interfaces with the new infrastructure. Can you look only at these parts?

File: packages/okta-adapter/src/filters/app_user_schema_deployment.ts [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+const makeModificationFromAddition = async (
+  change: AdditionChange<InstanceElement>,
+  client: OktaClient,
+  apiDefinitions: OktaSwaggerApiConfig,
+): Promise<ModificationChange<InstanceElement>> => {
+  const appUserSchemaInstance = getChangeData(change)
+  const parentApplicationId = getParents(appUserSchemaInstance)[0]?.id
+  if (parentApplicationId === undefined) {
+    log.error(`Error while trying to get parent id for user schema ${appUserSchemaInstance.elemID.getFullName()}`)
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/filters/app_user_schema_deployment.ts:112]: nit: there's another type called `user schema` so maybe change to `AppUserSchema` ? (same with the error below)

File: packages/okta-adapter/src/change_validators/app_user_schema_base_properties.ts [hunk 11]
```diff
@@ -23,53 +23,50 @@ import {
...
-  if (isAdditionChange(change)) {
-    return {
-      elemID: change.data.after.elemID,
-      severity: 'Warning',
-      message: 'Cannot deploy the base field in App User Schema',
-      detailedMessage:
-        'Okta API does not support deploying the base field in App User Schema. Salto will deploy the element without the base field. After fetch the field will be updated as the client',
-    }
-  }
+const getBasePath = (): string[] => [DEFINITIONS_FIELD, BASE_FIELD]
```
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/okta-adapter/src/change_validators/app_user_schema_base_properties.ts:33]: let's also export this and re-use in `app_user_schema_deployment` ?

File: packages/okta-adapter/src/filters/app_user_schema_deployment.ts [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+  autoCreatedappUserSchema: AppUserSchema,
+  appUserSchemaInstance: InstanceElement,
+  apiDefinitions: OktaSwaggerApiConfig,
+): InstanceElement => {
+  const createdAppUserSchemaInstance = appUserSchemaInstance.clone()
+  const { fieldsToOmit } = configUtils.getTypeTransformationConfig(
+    createdAppUserSchemaInstance.elemID.typeName,
+    apiDefinitions.types,
+    apiDefinitions.typeDefaults,
+  )
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/filters/app_user_schema_deployment.ts:96]: this part is deprecated, please use `fieldCustomizations` from `fetchDefinitions` instead.
there's an example for omitting hidden values [here](https://github.com/salto-io/salto/blob/main/packages/okta-adapter/e2e_test/adapter.test.ts#L466)
so it should be something like this but with `omit` instead of `hide`

File: packages/okta-adapter/src/filters/app_user_schema_deployment.ts [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+const AppUserSchemaSchema = Joi.object({
+  id: Joi.string().required(),
+}).unknown(true)
+
+const isAppUserSchema = (values: unknown): values is AppUserSchema => {
+  const { error } = AppUserSchemaSchema.validate(values)
+  return _.isUndefined(error)
+}
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/filters/app_user_schema_deployment.ts:54]: please use `createSchemeGuard`

**Review by shir-reifenberg — COMMENTED** [review]
Looks good 😎 
Didn't review the entire PR, added mainly nits and one comment about something that is probably not working after the fetch upgrade

**Review by shir-reifenberg — COMMENTED** [review]
adapter-components changes LGTM 😎 leaving the rest to @idoamit8

**Review by idoamit8 — APPROVED** [review]
LGTM! 💯 
NICLEY DONE 🍏
